Some tuple fixes / changes.

joelburget opened the pull request:
Hi Levent, this is a grab bag of a few tuple-related changes I've made on my end:

* When querying with a 0-tuple I encountered this error: `[FAIL] (declare-datatypes () ((() (()))))`. This clearly stems from `declSort` being called with `()`, though I'm a bit confused why we didn't encounter this before. Changing the `HasKind` / `SMTValue` instances for `()` fixed this for me.
* As part of that change I also made a special declaration for 0-tuples, `SBVTuple0`. I assume this could also be done in the same style as the other tuple sizes, but I slightly prefer declaring it as an enumeration.
* I needed a function `mkPair :: SBV a -> SBV b -> SBV (a, b)`. Really we should have this for all tuple sizes, though I haven't done this yet.
* I added a few tests I had written, in particular testing queries.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/leventerkok/sbv/442)
<!-- Reviewable:end -->

LeventErkok commented:
Isn't `mkPair` equivalent to `tuple`?

```
Prelude Data.SBV Data.SBV.Tuple> tuple (literal 'a', sTrue, 2) :: SBV (Char, Bool, Integer)
('a',True,2) :: (SChar, SBool, SInteger)
```

There's also `untuple` that goes the other way.

LeventErkok commented:
@joelburget I looked over this and it looks good; the 0-tuple case is indeed a nice catch.

I think `mkPair` is not needed, since we already have `tuple`/`untuple`. If you can remove that and update the patch; I'll pull this one in.

joelburget commented:
You're right -- `tuple` generalizes `mkPair` -- thanks :+1: 

I updated the PR:
* removed `mkPair`
* updated the guard line you noted
* updated one of my tests to use tuple
* added a fundep going the _other_ way on the `Tuple` class. this was necessary to get the previously mentioned test to compile without a type annotation. it agrees with my intuition that we can infer either of the parameters of `Tuple` from the other.
